src/devices/src/virtio/net/simple_handler.rs

```diff
@@ -0,0 +1,190 @@
+    // on the  metrics/events solution when they appear, and not propagate them further unless
+    // it's really useful/necessary.
+    fn write_frame_to_guest(&mut self) -> result::Result<bool, Error> {
+        let num_bytes = self.rxbuf_current;
+
+        let mut chain = match self.rxq.iter()?.next() {
+            Some(c) => c,
+            _ => return Ok(false),
+        };
+
+        let mut count = 0;
+        let buf = &mut self.rxbuf[..num_bytes];
+
+        while let Some(desc) = chain.next() {
+            let left = buf.len() - count;
+
+            if left == 0 {
+                break;
+            }
+
+            let len = cmp::min(left, desc.len() as usize);
+            chain
+                .memory()
+                .write_slice(&buf[count..count + len], desc.addr())
+                .map_err(Error::GuestMemory)?;
+
+            count += len;
+        }
+
+        if count != buf.len() {
+            // The frame was too large for the chain.
+            warn!("rx frame too large");
+        }
+
+        self.rxq.add_used(chain.head_index(), count as u32)?;
+
+        if self.rxq.needs_notification()? {
+            self.driver_notify.signal_used_queue(RXQ_INDEX);
+        }
+
+        self.rxbuf_current = 0;
+
+        Ok(true)
+    }
+
+    pub fn process_tap(&mut self) -> result::Result<(), Error> {
+        loop {
+            if self.rxbuf_current == 0 {
+                match self.tap.read(&mut self.rxbuf) {
+                    Ok(n) => self.rxbuf_current = n,
+                    Err(_) => {
+                        // Do something (logs, metrics, etc.) in response to an error when reading
+                        // from tap. EAGAIN means there's nothing available to read anymore (because
+                        // we open the TAP as non-blocking).
+                        return Ok(());
+                    }
+                }
+            }
+
+            if !self.write_frame_to_guest()? {
+                if !self.rxq.enable_notification()? {
+                    return Ok(());
+                }
+            }
+        }
+    }
+
+    fn send_frame_from_chain(
+        &mut self,
+        chain: &mut DescriptorChain<M>,
+    ) -> result::Result<u32, Error> {
+        let mut count = 0;
+
+        while let Some(desc) = chain.next() {
+            let left = self.txbuf.len() - count;
+            let len = desc.len() as usize;
+
+            if len > left {
+                warn!("tx frame too large");
+                break;
+            }
+
+            chain
+                .memory()
+                .read_slice(&mut self.txbuf[count..count + len], desc.addr())
+                .map_err(Error::GuestMemory)?;
+
+            count += len;
+        }
+
+        self.tap.write(&self.txbuf[..count]).map_err(Error::Tap)?;
+
+        Ok(count as u32)
+    }
+
+    pub fn process_txq(&mut self) -> result::Result<(), Error> {
```

Comment:
I was wandering if the TX can be done in a async fashion for performance reasons. Something like:
- do an async write on the tap without copying the packets from vm memory to the vmm;
- on the write completion notification mark the descriptor as used and signal the used queue.
Since this is tap device you shouldn't care of scatter gather.

Reply:
It's def helpful to skip as many copies as possible. We need to add some more support in various places to make that work safely & properly, but it's something we clearly want to do. If by async you also mean somehow submitting the requests to write/read data for each frame without waiting for them to complete in-between, then things become more nuanced. Overall, I imagine that doesn't provide a clear positive impact in this case, but would be interesting to explore at some point.